module/os/linux/zfs/zvol_os.c

```diff
@@ -542,84 +568,47 @@ zvol_open(struct block_device *bdev, fmode_t flag)
 		 * the kernel so the only option is to return the error for
 		 * the caller to handle it.
 		 */
-		if (!mutex_tryenter(&spa_namespace_lock)) {
-			rw_exit(&zvol_state_lock);
+		if (!mutex_owned(&spa_namespace_lock)) {
+			if (!mutex_tryenter(&spa_namespace_lock)) {
+				mutex_exit(&zv->zv_state_lock);
+				rw_exit(&zv->zv_suspend_lock);
 
 #ifdef HAVE_BLKDEV_GET_ERESTARTSYS
-			schedule();
-			return (SET_ERROR(-ERESTARTSYS));
-#else
-			if ((gethrtime() - start) > timeout)
+				schedule();
 				return (SET_ERROR(-ERESTARTSYS));
+#else
+				if ((gethrtime() - start) > timeout)
+					return (SET_ERROR(-ERESTARTSYS));
 
-			schedule_timeout(MSEC_TO_TICK(10));
-			goto retry;
+				schedule_timeout(MSEC_TO_TICK(10));
+				goto retry;
 #endif
-		} else {
-			drop_namespace = B_TRUE;
```

Comment:
I was going to say "I'm not familiar with this code, at all, but the change seems to match both the described problem and solution", but the CentOS testbots were [very](http://build.zfsonlinux.org/builders/CentOS%208%20x86_64%20%28TEST%29/builds/6389/steps/shell_4/logs/console) [unamused](http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/18121/steps/shell_4/logs/console) - and I'd speculate that the changes around only dropping it when this check passes are to blame. 

I'd probably try just guarding the mutex_exit with a mutex_owned check and see if CentOS screams less without breaking the world?

Reply:
Thanks for looking.  Yes, there was an issue here with dropping the namespace lock when we shouldn't.  Updated.